### PR TITLE
Update to work with Julia 1.5.x release

### LINKED ISSUE #1507: ClimateMachine.jl working with Julia 1.5.x release
# Description

There are currently GPU compilation errors with ClimateMachine on the new Julia 1.5 release (example build):
- https://buildkite.com/clima/climatemachine-ci/builds/185

@mwarusz has identified one KernalAbstraction issue with `eps` specialization:
https://github.com/JuliaGPU/KernelAbstractions.jl/issues/127

Suggested workaround of rewriting `eps(x)` to `x * eps(typeof(x))` breaks some of the numerical comparison tests:
- https://github.com/Clima/ClimateMachine.jl/commit/fe812a8a222fac6b5da0a20337f578089e781c70
- https://buildkite.com/clima/climatemachine-ci/builds/195

# For CLIMA Developers

- [ ] This feature can be added (if it cannot be, explain why in a comment below -- lack of technical expertise, not relevant to the scope of this project, too ambitious)
- [ ] There is a timeline for when this feature can be implemented
- [ ] The feature has been (or will be) implemented (Please link the PR)

### PR DESCRIPTION
**Pull request by jakebolewski**
# Description

Fix upstream and ClimateMachine GPU code generation issues, update to Julia 1.5

The Slurm CI needs to be switched over before being merged to use julia 1.5 by default (or better yet have an easier way to specify the version), but on the test configuration it is passing:

https://buildkite.com/clima/climatemachine-ci/builds/294

Closes #1507 

I have

- [ ] Written and run all necessary tests with CLIMA by including `tests/runtests.jl`
- [ ] Followed all necessary [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) and run `julia .dev/climaformat.jl .`
- [ ] Updated the documentation to reflect changes from this PR.

# For review by CLIMA developers

- [ ] There are no open pull requests for this already
- [ ] CLIMA developers with relevant expertise have been assigned to review this submission
- [ ] The code conforms to the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) and has consistent naming conventions. `julia .dev/format.jl` has been run in a separate commit.
- [ ] This code does what it is technically intended to do (all numerics make sense physically and/or computationally)

### CODE REVIEW

File: src/Land/Model/SoilWaterParameterizations.jl [hunk 1]
```diff
@@ -457,10 +457,7 @@ imaginary numbers, resulting in domain errors. Exit in this
 case with an error.
 """
 function effective_saturation(porosity::FT, ϑ_l::FT) where {FT}
-
-    if ϑ_l < 0
-        throw(DomainError(ϑ_l, "Effective saturation is negative."))
-    end
+    ϑ_l < 0 && error("Effective saturation is negative")
```
Comment (jakebolewski) [inline src/Land/Model/SoilWaterParameterizations.jl:460]: @kmdeck there is a codegen issue with throwing DomainError  (and other exception types) that our version of Cassette / KernelAbstractions and CUDA can't currently inline through in later Julia versions (1.5, 1.6).  I think we'll have to address what is going on on the codegen side but for now I switched to a less specific error which also matches much of the existing core numerics code running on the GPU.

**Comment by jakebolewski** [comment]
bors try

**Review by charleskawczynski — COMMENTED** [review]
I'd like to test this with #1308. Testing now... Looks like it worked! 🎉

**Review by charleskawczynski — APPROVED** [review]
LGTM

**Comment by jakebolewski** [comment]
@charleskawczynski I rebased the EDMF branch ontop of this and kicked off a build:
https://buildkite.com/clima/climatemachine-ci/builds/299

Looks like there a still some unit-test errors to work through.

**Comment by charleskawczynski** [comment]
> Looks like there a still some unit-test errors to work through.

It looks like this was done before some recent changes-- I see `julia --color=yes --project=docs/` in that , and I recently removed the plotting stuff so that this can run on the default env.

**Comment by charleskawczynski** [comment]
Maybe I did something wrong, I can try again today.

File: src/Land/Model/SoilWaterParameterizations.jl (hunk 1 above)
Comment (kmdeck) [reply src/Land/Model/SoilWaterParameterizations.jl:460]: hi! Thank you for taking care of this, and for updating the unit test. It's nice to know that a domain error on the variable \vartheta occurred (as a negative value here triggers other domain errors in certain functions, involving roots, e.g.) , but it looks like we get notified of that here still. looks good to me 👍

**Comment by jakebolewski** [comment]
@charleskawczynski with those fixes it is looking good:
https://buildkite.com/clima/climatemachine-ci/builds/300

The last issue is it seems like TicToc and some other methods are causing some invalidation errors when running with a pre-compiled system image.

**Comment by jakebolewski** [comment]
bors try

File: src/Arrays/MPIStateArrays.jl [hunk 2]
```diff
@@ -368,12 +368,12 @@ function begin_ghost_exchange!(Q::MPIStateArray; dependencies = nothing)
     __Irecv!(Q)
 
     # wait on (prior) MPI sends
-    @tic mpi_sendwait
+    # @tic mpi_sendwait
```
Comment (kpamnany) [inline src/Arrays/MPIStateArrays.jl:371]: Easier to just define empty macros in TicToc than to comment all these? Anyway, will hopefully find a solution today.

**Comment by jakebolewski** [comment]
bors try

**Comment by jakebolewski** [comment]
bors try

**Comment by jakebolewski** [comment]
@simonbyrne the final errors here seem to be due to differences between Julia versions when a `Random.seed!()` value is given in some of the unit tests.  Is it expected that a defined seed will give reproducible results across minor Julia versions (I would think so)?

**Comment by jakebolewski** [comment]
bors try

**Comment by jakebolewski** [comment]
bors try

**Comment by jakebolewski** [comment]
bors try

**Comment by jakebolewski** [comment]
bors try

**Pull request merged**